app/src/main/kotlin/io/homeassistant/companion/android/widgets/BaseGlanceEntityWidgetReceiver.kt

```diff
@@ -0,0 +1,284 @@
+
+            entitiesPerServer.forEach { (appWidgetId, entitiesPerServer) ->
+                widgetJobs[appWidgetId]?.cancel()
+
+                val serverId = entitiesPerServer.serverId
+                val entitiesId = entitiesPerServer.entityIds
+
+                val entityUpdatesFlow = if (serverManager.getServer(serverId) != null) {
+                    serverManager.integrationRepository(serverId).getEntityUpdates(entitiesId)
+                } else {
+                    null
+                }
+                if (entityUpdatesFlow != null) {
+                    widgetJobs[appWidgetId] = widgetScope.launch {
+                        Timber.tag(widgetClassName).d("Watching updates of entities($entitiesId) for widget $appWidgetId")
+                        entityUpdatesFlow.collect {
+                            onEntityUpdate(context, appWidgetId, it)
+                            updateView(context, appWidgetId)
+                        }
+                    }
+                } else {
+                    Timber.tag(widgetClassName).w("Entity updates is null for widget $appWidgetId not watching")
+                    // Shouldn't do anything since the job should have been canceled already
+                    removeSubscription(appWidgetId)
+                }
+            }
+        }
+    }
+
+    private fun stopWatchingForEntitiesChanges() {
+        Timber.tag(widgetClassName).d("Stop watching.")
+        widgetScope.cancel()
+        widgetJobs.clear()
+    }
+
+    private fun removeSubscription(appWidgetId: Int) {
+        Timber.tag(widgetClassName).d("Removing subscription for widget = $appWidgetId")
+        widgetJobs.remove(appWidgetId)?.cancel()
+    }
+
+    private suspend fun cleanupOrphansWidgetAndGetEntitiesByServer(
+        context: Context,
+    ): Map<Int, EntitiesPerServer> {
+        val manager = glanceManagerProvider(context)
+        val glanceIds = manager.getGlanceIds(glanceAppWidget.javaClass)
+
+        val systemWidgetIds = glanceIds.map { manager.getAppWidgetId(it) }
+
+        val entitiesPerServer = getWidgetEntitiesByServer(context)
+
+        val invalidWidgetIds = entitiesPerServer.keys.minus(systemWidgetIds)
+
+        if (invalidWidgetIds.isNotEmpty()) {
+            Timber.tag(widgetClassName).i(
+                "Found widgets $invalidWidgetIds in database, but not in AppWidgetManager - removing it from the database.",
+            )
+            deleteWidgetsFromDatabase(invalidWidgetIds.toIntArray())
+        }
+
+        // strip the server and entities from the map since they are not used anymore
+        val entitiesPerServerWithoutInvalid = entitiesPerServer - invalidWidgetIds
+
+        return entitiesPerServerWithoutInvalid
+    }
+
+    /**
+     * From https://cs.android.com/androidx/platform/frameworks/support/+/androidx-main:glance/glance-appwidget/src/main/java/androidx/glance/appwidget/GlanceAppWidget.kt;l=78;drc=852af4153b5fc3f0c92b8356f00ec5d786c5e869
```

Comment:
- Can you clean up the url? `;drc=852af4153b5fc3f0c92b8356f00ec5d786c5e869` isn't necessary.
- Can you permalink the url? `main` will change in the future.

Reply:
I need to check how to do it from cs.android 👍

Reply:
Click the arrow next to the branch name to switch to the latest commit, or View in > View in Gitiles
![image](https://github.com/user-attachments/assets/8aa48538-00ea-44f2-943a-e52e7d989f74)

Reply:
Done, but I'm curious do you what is DRC ?

Reply:
Sorry, no idea what DRC is. Probably some tracking/UI state that isn't relevant to us.